Clean up some code around m_objectBoundingBox

smfr opened the pull request:
#### bb9b920961a04bd6e58bbe067d9bdae6f74d5bc7
<pre>
Clean up some code around m_objectBoundingBox
<a href="https://bugs.webkit.org/show_bug.cgi?id=291721">https://bugs.webkit.org/show_bug.cgi?id=291721</a>
<a href="https://rdar.apple.com/149514038">rdar://149514038</a>

Reviewed by Said Abou-Hallawa.

LegacyRenderSVGRoot and LegacyRenderSVGContainer have similar behavior where they store
a `FloatRect m_objectBoundingBox`, and separate `bool m_objectBoundingBoxValid`.

Convert those to use Markable&lt;FloatRect&gt;, choosing Markable&lt;&gt; to match the other data members.
There were code paths that could use m_objectBoundingBox when m_objectBoundingBoxValid was
false; update them to use `value_or()`.

Change `SVGRenderSupport::computeContainerBoundingBoxes()` to return a struct with the two
rects, rather than using reference arguments.

Note that in `updateCachedBoundaries()`, we have to set `m_objectBoundingBox` before
calling `intersectRepaintRectWithResources()` because in some cases, resource bounds
are effected by `m_objectBoundingBox` (exercised by svg/filters/css-repaint-reference-filter-on-root.html).

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
(WebCore::updateObjectBoundingBox): Deleted. Moved inline for clarity.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::updateCachedBoundaries):
(WebCore::LegacyRenderSVGContainer::repaintRectInLocalCoordinates const):
(WebCore::LegacyRenderSVGContainer::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
(WebCore::LegacyRenderSVGContainer::isObjectBoundingBoxValid const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::LegacyRenderSVGRoot):
(WebCore::LegacyRenderSVGRoot::updateCachedBoundaries):
(WebCore::LegacyRenderSVGRoot::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:

Canonical link: <a href="https://commits.webkit.org/293861@main">https://commits.webkit.org/293861@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5613fda5641daf571bc8e76eb6790e459117ac27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100071 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19719 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10012 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105199 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50652 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102112 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20025 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28192 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76172 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33245 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15293 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90375 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56533 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15108 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8371 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50021 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85019 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107559 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27184 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19906 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86581 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84661 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21521 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29337 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7070 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21020 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27121 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32350 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26932 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30248 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28491 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
```diff
@@ -247,7 +248,7 @@ bool LegacyRenderSVGContainer::nodeAtFloatPoint(const HitTestRequest& request, H
 
     // Accessibility wants to return SVG containers, if appropriate.
-    if (request.type() & HitTestRequest::Type::AccessibilityHitTest && m_objectBoundingBox.contains(localPoint)) {
+    if (request.type() & HitTestRequest::Type::AccessibilityHitTest && objectBoundingBox().contains(localPoint)) {
```
Was it a bug that we do not check `m_objectBoundingBoxValid` before using `m_objectBoundingBox`? Or do we need to `ASSERT(static_cast<bool>(m_objectBoundingBox))` here?

smfr replied:
It was a bug, yes.

shallawa commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -138,28 +138,26 @@ LayoutRepainter::CheckForRepaint SVGRenderSupport::checkForSVGRepaintDuringLayou
-    if (!otherValid)
-        return;
+    // Update a bounding box taking into account the validity of the other bounding box.
+    auto updateObjectBoundingBox = [&](const RenderObject* other, FloatRect otherBoundingBox) {
```
It is not clear to me what this lambda is doing. I think it unites two boundingBoxes. But it is called once in this function now. Do we still to have this code as a lambda?

smfr replied:
It's called per child.

shallawa commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -178,13 +176,17 @@ void SVGRenderSupport::computeContainerBoundingBoxes(const RenderElement& contai
+            repaintRect = transform.mapRect(repaintRect);
         }
+
+        updateObjectBoundingBox(current.ptr(), objectBounds);
```
I think this code can be simplified like this: 
```
CheckedPtr container = dynamicDowncast<LegacyRenderSVGContainer>(current->ptr());
if (container->isObjectBoundingBoxValid()) {
    if (!objectBoundingBox)
        objectBoundingBox = objectBounds;
    else
        objectBoundingBox->uniteEvenIfEmpty(objectBounds);
}
```

shallawa commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -178,13 +176,17 @@ void SVGRenderSupport::computeContainerBoundingBoxes(const RenderElement& contai
+        auto repaintRect = current->repaintRectInLocalCoordinates(repaintRectCalculation);
+
+        if (!transform.isIdentity()) {
+            objectBounds = transform.mapRect(objectBounds);
```
Maybe we need to check `isObjectBoundingBoxValid()` be doing this transform.

smfr committed:
Clean up some code around m_objectBoundingBox
https://bugs.webkit.org/show_bug.cgi?id=291721
rdar://149514038

Reviewed by Said Abou-Hallawa.

LegacyRenderSVGRoot and LegacyRenderSVGContainer have similar behavior where they store
a `FloatRect m_objectBoundingBox`, and separate `bool m_objectBoundingBoxValid`.

Convert those to use Markable<FloatRect>, choosing Markable<> to match the other data members.
There were code paths that could use m_objectBoundingBox when m_objectBoundingBoxValid was
false; update them to use `value_or()`.

Change `SVGRenderSupport::computeContainerBoundingBoxes()` to return a struct with the two
rects, rather than using reference arguments.

Note that in `updateCachedBoundaries()`, we have to set `m_objectBoundingBox` before
calling `intersectRepaintRectWithResources()` because in some cases, resource bounds
are effected by `m_objectBoundingBox` (exercised by svg/filters/css-repaint-reference-filter-on-root.html).

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
(WebCore::updateObjectBoundingBox): Deleted. Moved inline for clarity.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::updateCachedBoundaries):
(WebCore::LegacyRenderSVGContainer::repaintRectInLocalCoordinates const):
(WebCore::LegacyRenderSVGContainer::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
(WebCore::LegacyRenderSVGContainer::isObjectBoundingBoxValid const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::LegacyRenderSVGRoot):
(WebCore::LegacyRenderSVGRoot::updateCachedBoundaries):
(WebCore::LegacyRenderSVGRoot::repaintRectInLocalCoordinates const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:

Canonical link: https://commits.webkit.org/293861@main
